### ISSUE TITLE
Performance comparisons between bulkPut and put?

### ISSUE THREAD

**Issue by Zloka** [dfahlander/Dexie.js#1427]
I am using Dexie in an offline first web app. To get the same experience offline, the web app performs a periodic (or otherwise explicitly triggered) check for missing dependencies, and downloads the dependencies ahead of when the user might need it.

Long story short, the missing data is mainly binary files, more specifically PDF files, and are typically ~50-100kb in size, though exceptions with larger files exist. The table that saves these files have no indices, as dynamically construct the key and the binary data is the only thing stored, which I won't query.

The current logic is essentially:
1. Determine a list of missing dependencies.
2. Fetch an individual missing dependency, and persist it offline using Dexie's `put` method.

However, I was wondering if this approach might be more sensible:
1. Determine a list of missing dependencies.
2. Fetch `N` missing dependencies (for example 30, or whatever makes sense given the average size of my dependencies).
3. Persist `N` dependencies offline using Dexie's `bulkPut` method.

I'm sorry if this has been asked before. Based on the documentation, it seems that using `bulkPut` over `put` should speed things up at least for a large number `N`. In my case, a realistic amount of dependencies is <10000, which doesn't seem that large but on the other had I am storing binary data, which may or may not be relevant.

If there are some performance comparisons I have failed to find, or some other knowledge with regards to how Dexie might perform in this particular case, I'm all ears!

**Comment by dfahlander** [comment]
Hi! bulkPut is faster but you might need to divide your calls into chunks to avoid overloading IndexedDB if files are large, I suppose. Using bulkPut in chunks of 10 is still faster than just using put on each. I don't have any numbers on what would be considered a good chunk size so you'd have to test it out. Based on your files being 50-100kb each I would guess it to be ok with a chunk size of 100.

**Comment by FunMiles** [comment]
Just a practical comment: 

- Won't a `fetch(fileURL)` take way longer than the time needed for a single `put`?
- The `put` can happen while other `fetch` operations are going on.
- After the `put`, the variable holding the PDF will run out of scope, allowing the GC to get rid of the memory space, reducing the app memory usage.
Given those remarks, unless someone see something I missed, I would't bother with a `bulkPut`.

**Comment by Zloka** [comment]
Thank you for your comments!

Yeah, I oversimplified my initial approach a little bit, essentially I have `fetchAndPut` promises, and run up to 8 of those concurrently.

I think I'll try to prototype if separating the `fetch` and `put` calls is an improvement. Perhaps I could run more `fetches` concurrently, and `bulkPut` a bunch of them when I have fetched enough dependencies. If I get some useful comparisons I'll post them here, but feel free to close this issue for housekeeping if you wish! 🙂
